Use on-the-fly rewrite (no more file writes)

dhanak opened the pull request:
I really like this package, but IMHO the rewriting of the extension packages could be done without writing to the file system, providing a more robust approach. Here's my take on it.

dhanak committed:
Use on-the-fly rewrite (no more file writes)

pat-alt commented:
I also really like this package and I like this PR.

I think that avoiding the file write could help with the fact that currently, you cannot split and organise the files in `ext` into various subfiles. I originally did this in this [PR](https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/pull/270):

```
ext
├── MPIExt
│   ├── MPIExt.jl
│   └── utils.jl
├── PythonCallExt
│   ├── PythonCallExt.jl
│   ├── models.jl
│   └── utils.jl
└── RCallExt
    ├── RCallExt.jl
    └── utils.jl
```

Using this setup tests failed on Julia versions `<1.9` after [warnings](https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/actions/runs/6026392570/job/16349210977#step:6:1506) such as the one below that certain extension files couldn't be located:

```julia
┌ Warning: Error requiring `MPI` from `CounterfactualExplanations`
│   exception =
│    LoadError: SystemError: opening file "/home/runner/work/CounterfactualExplanations.jl/CounterfactualExplanations.jl/ext_compat/MPIExt/utils.jl": No such file or directory
│    Stacktrace:
[...]
```

Eventually, I [moved](https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/pull/270/commits/139aae01dda74be059b4ea6fa29a5a9ed267cc4f) all code into single files

```
ext
├── MPIExt.jl
├── PythonCallExt.jl
└── RCallExt.jl
```

and all tests passed, but the previous folder structure was more organised.

cjdoris commented:
@dhanak Thanks for the PR. This is actually how an early version (pre v1) of the package worked, but the 2-argument version of `include` was introduced in Julia 1.5 - hence the tests are failing on earlier versions. Since this is a compat package, I wanted it to work all the way back to Julia 1.0.

So there are a few options:
1. The current approach: rewrite the code to disk.
2. Only support Julia 1.5+ and rewrite on the fly.
3. Use both strategies depending on the Julia version.

I'd like to avoid 3 since it's more code to support and test - and might introduce subtle differences in behaviour between versions.

Arguably option 2 is ok since Julia 1.6 is LTS. It would be informative to look at the Julia bound in packages using PackageExtensionCompat.

cjdoris commented:
@pat-alt Your issue is that only the root file for each extension gets rewritten currently. It would be straightforward to copy all the ext code over - could you make a separate issue for that please?

dhanak commented:
Yes, now I can see from the tests that the two parameter include doesn't work before Julia 1.5. My mistake, but still a pity.

Perhaps this on-the-fly variant of the `PackageExtensionCompat` can be versioned 2.0, and it would be compatible with Julia versions 1.5 (or 1.6 LTS) up only. The pre-1.5 Julia support could still be supported by the 1.0 branch. I doubt that a lot of maintenance will be required for this code base. This way, the package manager would automatically pick up the best version for any Julia release.

WDYT?

cjdoris commented:
Yeah I considered that option, but I don't think it's really any different from option 3 above (branch on the Julia version) - the end result is the same and the amount of code to support is about the same. Personally I'd prefer option 3 - since then at least all the code is in one place.

A disadvantage of your option 4 is that somebody may reasonably put a compat bound `PackageExtensionCompat = "2"` in their package and now it won't work on Julia 1.0.

Is there any real objection to the status quo other than that icky feeling you get from rewriting code to disk just to immediately read it in again?

dhanak commented:
> Yeah I considered that option, but I don't think it's really any different from option 3 above (branch on the Julia version) - the end result is the same and the amount of code to support is about the same. Personally I'd prefer option 3 - since then at least all the code is in one place.

A slight benefit would be that the code itself would be clearer in either variant, but you have a fair point.

> A disadvantage of your option 4 is that somebody may reasonably put a compat bound `PackageExtensionCompat = "2"` in their package and now it won't work on Julia 1.0.

This is all hypothesizing, but IMHO, if one were to maintain backward compatibility with Julia 1.0, one would surely check the package versions installed by Julia 1.0, where it would be immediately clear that PackageExtensionCompat v1 is being used.
 
> Is there any real objection to the status quo other than that icky feeling you get from rewriting code to disk just to immediately read it in again?

For one, you need to add `ext_compat` to `.gitignore`. And at least on one occasion, when I tested my package with various Julia versions and deleted the `ext_compat` directory several times in the process, at one point, it just complained about it missing, but didn't regenerate it. I don't know what the exact steps were, and I think it got resolved by deleting the `Manifest.toml` file as well and then re-resolving the dependencies. It was just a minor nuisance.

So overall, I don't really mind if everything stays as is, and it's your call to make, anyhow.

dhanak commented:
That being said, I came up with a workaround that makes the parsing solution viable also for pre 1.5 versions. See the updated PR, please.

dhanak committed:
Make parser work with pre-1.5 Julia versions as well

cjdoris commented:
IIUC this PR now implements the 3-arg version of `include` itself? Does it work if the extension has more than one source file (i.e. if the main extension file includes another one)? The tests don't cover this case but I suspect the inner includes will break, because the relative file names won't be resolvable. (Admittedly the current version also has this limitation but it's easy to fix.)

cjdoris commented:
> This is all hypothesizing, but IMHO, if one were to maintain backward compatibility with Julia 1.0, one would surely check the package versions installed by Julia 1.0, where it would be immediately clear that PackageExtensionCompat v1 is being used.

I think you're right, this particular objection would be caught by proper testing.

> For one, you need to add `ext_compat` to `.gitignore`. And at least on one occasion, when I tested my package with various Julia versions and deleted the `ext_compat` directory several times in the process, at one point, it just complained about it missing, but didn't regenerate it. I don't know what the exact steps were, and I think it got resolved by deleting the `Manifest.toml` file as well and then re-resolving the dependencies. It was just a minor nuisance.

How about this package also writes out a `ext_compat/.gitignore` to ignore everything? Should solve your first issue. If your second issue (missing extension code) is reproducible please file an issue.

> So overall, I don't really mind if everything stays as is, and it's your call to make, anyhow.

I'm leaning this way - the current approach is nice and simple. I'd certainly prefer doing the transformation in-memory but without the 2-arg `include` I think it'll be hard to get right in general (we'd be reimplementing `include` ourselves).

dhanak committed:
Handle inner includes and extension submodules correctly

The `_include` and `rewrite` functions needed a bit of tweaking to
handle inner includes and extension submodules correctly. Also added two
tests for these scenarios.

dhanak commented:
> IIUC this PR now implements the 3-arg version of `include` itself? Does it work if the extension has more than one source file (i.e. if the main extension file includes another one)? The tests don't cover this case but I suspect the inner includes will break, because the relative file names won't be resolvable. (Admittedly the current version also has this limitation but it's easy to fix.)

You are perfectly right. It doesn't take care of inner includes, just as the current version, as you point it out. But this can be taken care of. While fixing this, I also noticed that neither implementation works well if the extension has a submodule and makes use of the dependent package from there. The `use ..WeakDep` form doesn't work in this case, because the `..` is insufficient.

I fixed all these in the PR, please see the updates.

cjdoris commented on src/PackageExtensionCompat.jl:
```diff
@@ -5,39 +5,55 @@ export @require_extensions
+            cd(dirname(path)) do
+                str = read(basename(path), String)
+                pos = 1
+                while true
```
The file should contain a single expression (the module implementing the extension) so you don't need a loop - can just `expr = Meta.parse(str)`.

dhanak replied:
For the extension module, I agree that it should. Is there anything, though, that prevents the developer to put more than one block in a file? Also, this same `_include` function will be used to load additional files, e.g., files implementing parts of the extension module, and it is perfectly legal to have multiple definitions in those files. So I think that the for loop cannot be omitted.

cjdoris replied:
I assume an exception is thrown if the top level file doesn't contain exactly a single `module WeakDep ... end` expression. And I don't think you actually use `_import` anywhere but the top level? Any imports in the extension code are left as-is.

cjdoris replied:
Still if it works it works and I'm happy to leave it.

dhanak replied:
Not quite, _all_ includes within the extension package have to be rewritten recursively, in order to find all references to the `WeakDep`s. So the same `_include` implementation will be used for nested includes as well, which, as I said, don't have to limit themselves to a single module definition. (Even though the unit test does that.)

cjdoris commented on src/PackageExtensionCompat.jl:
```diff
@@ -5,39 +5,55 @@ export @require_extensions
-        parts = map(part -> rewrite_import(part, pkgs), parts)
-        join(parts, ", ")
+    function rewrite(toplevel::Module, pkgs)
+        Base.Fix1(MacroTools.postwalk, block -> rewrite_block(block, toplevel, pkgs))
```
I think this is the only place MacroTools is used? Can you remove it? Trying to keep dependencies as minimal as possible.

dhanak replied:
I'll look into it.

cjdoris commented on src/PackageExtensionCompat.jl:
```diff
@@ -5,39 +5,55 @@ export @require_extensions
-        write(trgfile, code)
+    function rewrite_use(use::Expr, toplevel::Module, pkgs)::Expr
+        @assert Meta.isexpr(use, :.)
+        string(use.args[1]) ∈ pkgs ? Expr(:., nameof(toplevel), use.args...) : use
```
This rewrites `using WeakDep` to `using TopPkg.WeakDep`? Can you leave a comment to that effect?

I was a bit worried this might not be as robust as `using ..WeakDep` (i.e. what if `using TopPkg` doesn't work) but I think it's actually fine because the `TopPkg` *must* be an installed package in order to use package extensions, and therefore `using TopPkg` will always succeed. Is this right? Any cases where this wouldn't work?

dhanak replied:
Yes, you figured that out correctly. I'll add a comment.

After giving this some thought, I now think that `TopPkg.WeakDep` is a safer way of referring to the dependency than `..WeakDep`, because it works from any level of the module hierarchy.

I can think of a somewhat forced scenario where this possibly wouldn't work:

```
module TopPkg
   module SubPkg
       const TopPkg = nothing
       # wouldn't work here since TopPkg doesn't refer to the package locally
       using WeakDep
   end
end
```

But I wouldn't expect anyone in their right mind to hide `TopPkg` like that.

Other than that, `TopPkg` should be reachable using the package manager, AFAIK.

cjdoris replied:
Do you mean
```julia
const TopPkg = nothing
using TopPkg.WeakDep
```
?

I think even this will work. AFAIU `using X` always imports the package `X` regardless of the local scope. It's only things like `using .X` which depend on the scope. So in conclusion I think we're good.

dhanak replied:
```
using TopPkg.WeakDep
```

That's what it would be rewritten to, correct. And you are probably right about `using` ignoring the local scope.

cjdoris reviewed (changes requested):
Thanks for adding tests for submodules and includes. I'm now quite in favour of this PR. There's a few comments to address.

dhanak committed:
Replace `MacroTools` with local implementation of `postwalk`

Also add some explanatory comments.

cjdoris commented on Project.toml:
```diff
@@ -1,7 +1,7 @@
 uuid = "65ce6f38-6b18-4e1d-a461-8949797d7930"
 authors = ["Christopher Doris <github.com/cjdoris>"]
-version = "1.0.1"
+version = "1.1.0"
```
1.0.2? This PR doesn't change the API.

dhanak replied:
Whatever you say, this is your package :-). I thought that perhaps this isn't just a simple patch, and therefore at least a minor version increase would be due to indicate that, even though there is no change in the API. But if you would be more comfortable with a patch version increase, than that's whay you get.

dhanak committed:
Increase Patch version instead of minor version

The API hasn't changed, so minor version doesn't need to change, either.

cjdoris commented on src/PackageExtensionCompat.jl:
```diff
@@ -7,37 +7,63 @@ const HAS_NATIVE_EXTENSIONS = isdefined(Base, :get_extension)
-            line
+    function rewrite_block(block, top_pkg::Module, pkgs)
+        if Meta.isexpr(block, :call) && block.args[1] == :include
+            # inner include, rewrite it recursively
```
I take back an earlier comment - you do rewrite includes recursively. You should check it's being called with one argument.

dhanak replied:
Up to julia 1.4, there is only one method in the global `include` function, and that method takes a single string as an argument. So I doubt there will be any other kind of call.

From julia 1.5, there is also a two argument variant, but the same version introduces the three arugment method of `Base.include`, and we will use that.

But if you still want this argcount check, I can add it for you.

cjdoris replied:
Yeah please check.

dhanak replied:
Sorry, I made this change 5 days ago, just forgot to push it. Duh.

dhanak committed:
Check `include` call parameter count before replacing

cjdoris commented:
Thanks very much for this!
